File: toris_collection/flock.py
"""
flock.py - 群れ(同種が複数)のサイズを決める純粋計算

■ 方針(交渉不能・HANDOFF §1-1 の背骨に沿う)
  群れは「ラジオを豊かにする」装置。新種を増やさずに、馴染んだ鳥の声を
  厚くする。よく会った鳥ほど大きな群れで鳴く = 「会いに行く」がそのまま
  声の厚みに accrete する。これは軸(ラジオが豊かになる / また会いに行く)
  に直結する。罰でも損でもなく、ただ豊かさが積み上がる。

■ 生態学的根拠(誠実さ)
  - 群れ形成は種に強く依存する: 小型で普通種(スズメ・メジロ等)は群れ、
    猛禽や縄張り性の強い種は単独でいることが多い。
    ここでは rarity を素朴な代理変数として使う(普通種ほど群れやすい)。
    種ごとに正確に決めたいときは Sheets の flock_max 列で上書きできる。
  - 群れの大きさは生息地の質と馴染みで育つ。本モジュールは観察回数
    (familiarity)で cap まで成長させる。馴染みのない鳥はまず1羽から。

このモジュールは I/O にも UI にも依存しない純粋計算。
"""
from __future__ import annotations

MAX_CAP = 3          # 1種あたりの群れ上限(音が濁らない範囲に抑える)
GROWTH_EVERY = 3     # 観察何回ごとに群れが1羽増えるか


def flock_cap(bird_id: str, birds_data: dict) -> int:
    """その種が作りうる群れの最大サイズ(1..MAX_CAP)。

    data に flock_max があればそれを使う(Sheets 列で種ごとに調整可)。
    無ければ rarity から導く: 普通種=群れやすい / レア種=単独。
    """
    bird = birds_data.get(bird_id, {})
    explicit = bird.get("flock_max")
    if explicit is not None:
        try:
            return max(1, min(MAX_CAP, int(explicit)))
        except (TypeError, ValueError):
            pass
    try:
        r = float(bird.get("rarity", 0.5))
    except (TypeError, ValueError):
        r = 0.5
    if r >= 0.7:
        return 1      # レア = 単独(特別な一声)
    if r >= 0.4:
        return 2
    return 3          # 普通種 = 群れやすい


def flock_size(bird_id: str, count: int, birds_data: dict) -> int:
    """ラジオで今この鳥が何羽で鳴くか(1..cap)。

    cap を超えない範囲で、観察回数(familiarity)に応じて群れが育つ。
    単独種(cap=1)は常に1。よく会った社会性の鳥ほど厚い群れになる。
    """
    cap = flock_cap(bird_id, birds_data)
    if cap <= 1:
        return 1
    try:
        c = int(count)
    except (TypeError, ValueError):
        c = 0
    grown = 1 + max(0, c) // GROWTH_EVERY
    return max(1, min(cap, grown))

File: toris_core/lib/src/flock.dart
/// 群れ(同種が複数)のサイズを決める純粋計算。`toris_collection/flock.py` の移植。
///
/// 群れは「ラジオを豊かにする」装置。新種を増やさずに、馴染んだ鳥の声を厚くする。
/// よく会った鳥ほど大きな群れで鳴く = 「会いに行く」がそのまま声の厚みになる。
///
/// 生態学的な根拠は Python 版の docstring を参照(群れやすさは種に依存し、
/// ここでは rarity を素朴な代理変数として使う)。
library;

import 'py_coerce.dart';

/// 1種あたりの群れ上限(音が濁らない範囲に抑える)。
const int maxCap = 3;

/// 観察何回ごとに群れが1羽増えるか。
const int growthEvery = 3;

/// その種が作りうる群れの最大サイズ(1..maxCap)。
///
/// データに flock_max があればそれを使う(種ごとの上書き)。無ければ rarity から
/// 導く: 普通種=群れやすい / レア種=単独。
int flockCap(String birdId, Map<String, dynamic> birdsData) {
  final bird = (birdsData[birdId] as Map?) ?? const {};
  final explicit = bird['flock_max'];
  if (explicit != null) {
    final n = pyInt(explicit);
    // Python 版は int() が失敗したときだけ rarity へ落ちる。数として読めたなら
    // その値を 1..maxCap に丸めて使う。
    if (n != null) return n.clamp(1, maxCap);
  }
  final r = pyFloat(bird['rarity']) ?? 0.5;
  if (r >= 0.7) return 1; // レア = 単独(特別な一声)
  if (r >= 0.4) return 2;
  return 3; // 普通種 = 群れやすい
}

/// ラジオで今この鳥が何羽で鳴くか(1..cap)。
///
/// cap を超えない範囲で、観察回数に応じて群れが育つ。単独種(cap=1)は常に1。
int flockSize(String birdId, Object? count, Map<String, dynamic> birdsData) {
  final cap = flockCap(birdId, birdsData);
  if (cap <= 1) return 1;
  final c = pyInt(count) ?? 0;
  // Python の `max(0, c) // GROWTH_EVERY`(切り捨て除算)
  final grown = 1 + (c < 0 ? 0 : c) ~/ growthEvery;
  return grown.clamp(1, cap);
}
